OADP- 2300 OADP Metric Collection (#62710)
openshift/openshift-docs: modules/oadp-monitoring.adoc

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/troubleshooting.adoc

:_content-type: CONCEPT
[id="oadp-monitoring_{context}"]

> = OADP Monitoring

Author: I think it should be viewed as a component of the OADP.

> The {product-title} provides a xref:../../monitoring/monitoring-overview.adoc#about-openshift-monitoring[monitoring stack] that allows users and administrators to effectively monitor and manage their OpenShift clusters, as well as monitor and analyze the workload performance of user applications and services running on the clusters, including receiving alerts if an event occurs.

Reviewer: Is the `{product-title}` here replaced by "OADP Monitoring" If yes I think it's not precise.

OADP Monitoring does not provide monitoring stack, it *uses* it.

Author: The {product-title} is OADP.

Author: Oops it's not OADP.

OCP provides a monitoring stack...

  labels:
    app: oadp-service-monitor
  name: oadp-service-monitor
  namespace: openshift-adp
spec:
  endpoints:
  - interval: 30s
    path: /metrics
    targetPort: 8085
    scheme: http
  selector:
    matchLabels:
      app.kubernetes.io/name: "velero"
----
+
. Apply the `3_create_oadp_service_monitor.yaml`.
+
[source,terminal]
----
$ oc apply -f 3_create_oadp_service_monitor.yaml
servicemonitor.monitoring.coreos.com/oadp-service-monitor created
----
+
. Confirm that the new ServiceMonitor is `Up`.
In the *Administrator* perspective of the OpenShift Container Platform web console use *Observe/Targets* to view the Metrics Targets. Ensure the `Filter` is unselected or that the `User` source is selected, and type `openshift-adp` in the `Text` search field. Ensure the status for the ServiceMonitor is `Up`.
+
.OADP metrics targets

image::oadp-metrics-targets.png[OADP metrics targets]

[id="Sample-alerting-rules_{context}"]
==  Sample alerting rules

The OpenShift Container Platform monitoring stack allows to receive Alerts configured using Alerting Rules. To create an Alerting rule for the OADP project, use one of the Metrics which are scraped with the user workload monitoring.

> Refer to xref:../../monitoring/managing-alerts.adoc#managing-alerts[managing alerts] in the OpenShift documentation for detailed instructions on how to create and manage OpenShift Alerts.

Reviewer 2: x2 "OpenShift" - can the first be `{product-title}` and maybe just drop the second one?

+
[source,yaml]
+
----
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: sample-oadp-alert
  namespace: openshift-adp
spec:
  groups:
  - name: sample-oadp-backup-alert
    rules:
    - alert: OADPBackupFailing
      annotations:
        description: 'OADP had {{$value | humanize}} backup failures over the last 2 hours.'
        summary: OADP has issues creating backups
      expr: |
        increase(velero_backup_failure_total{job="openshift-adp-velero-metrics-svc"}[2h]) > 0
      for: 5m
      labels:
        severity: warning
----
+
. Apply the `4_create_oadp_alert_rule.yaml`, this creates the PrometheusRule in the `openshift-adp namespace`.
+
[source,terminal]
----
$ oc apply -f 4_create_oadp_alert_rule.yaml
prometheusrule.monitoring.coreos.com/sample-oadp-alert created
----
+
. After the Alert is triggered, it can be viewed in the following ways:
.. In the *Administrator* view under *Observe/Alerting* menu. You must select `User` in the `Filter` box, otherwise by default only the `Platform` alerts are displayed.

> .. In the *Developer* view, select the *Observe* menu.

Reviewer 2: These substeps are currently numbered, but seem like they could be unordered steps (asterisks for bullets), since you can do either / they're not sequential?

You could also consider turning this last step into a `.Verification` section (["Procedure verification"](https://redhat-documentation.github.io/modular-docs/#con-creating-procedure-modules_writing-mod-docs)) for the procedure.

> .. In the *Administrator* view under *Observe/Alerting* menu. You must select `User` in the `Filter` box, otherwise by default only the `Platform` alerts are displayed.
> .. In the *Developer* view, select the *Observe* menu.

Reviewer 2: ```suggestion
.. In the *Administrator* view under *Observe* -> *Alerting* menu, select *User* in the *Filter* box. Otherwise, by default only the *Platform* alerts are displayed.
.. In the *Developer* view, select the *Observe* menu.
```

+
.OADP backup failing alert

image::oadp-backup-failing-alert.png[OADP backup failing alert]

> [id="List-of-metrics_{context}"]
> ==  List of available metrics

Reviewer 2: ```suggestion
[id="list-of-metrics_{context}"]
== List of available metrics
```

^ Lowercase only in `id`s, also extra space in heading.

Altho, could the `id` be a little more specific? E.g., `[id="oadp-metrics-list_{context}"]`

Reviewer 2: Also, break this out into a reference module.

These are the list of metrics provided by the OADP together with their https://prometheus.io/docs/concepts/metric_types/[Types].

|===

> |Metric Name |Description |Type

Reviewer 2: ```suggestion
|Metric name |Description |Type
```

> |kopia_content_cache_hit_bytes

Reviewer 2: I feel like each of these "metric name" entries should be marked up, probably backticks.

|Number of bytes retrieved from the cache
|Counter

|kopia_content_cache_hit_count
|Number of time content was retrieved from the cache
|Counter

|kopia_content_cache_malformed
|Number of times malformed content was read from the cache
|Counter

|kopia_content_cache_miss_count
|Number of time content was not found in the cache and fetched
|Counter

|kopia_content_cache_missed_bytes
|Number of bytes retrieved from the underlying storage
|Counter

|kopia_content_cache_miss_error_count
|Number of time content could not be found in the underlying storage
|Counter

|kopia_content_cache_store_error_count
|Number of time content could not be saved in the cache
|Counter

|kopia_content_get_bytes
|Number of bytes retrieved using GetContent
|Counter

|kopia_content_get_count

> |Number of time GetContent() was called

Reviewer 2: ```suggestion
|Number of times `GetContent()` was called
```

or

```suggestion
|Number of times the `GetContent()` function was called
```

Don't know if "function" would be the right descriptor here.

Same suggestions for the other `GetContent()` and `WriteContent()` entries below.

|Counter

|kopia_content_get_error_count
|Number of time GetContent() was called and the result was an error
|Counter

|kopia_content_get_not_found_count
|Number of time GetContent() was called and the result was not found
|Counter

|kopia_content_write_bytes
|Number of bytes passed to WriteContent()
|Counter

|kopia_content_write_count
|Number of time WriteContent() was called
|Counter

|velero_backup_attempt_total
|Total number of attempted backups
|Counter

|velero_backup_deletion_attempt_total
|Total number of attempted backup deletions
|Counter

|velero_backup_deletion_failure_total
|Total number of failed backup deletions
|Counter

|velero_backup_deletion_success_total
|Total number of successful backup deletions
|Counter

|velero_backup_duration_seconds
|Time taken to complete backup, in seconds
|Histogram

|velero_backup_failure_total
|Total number of failed backups
|Counter

|velero_backup_items_errors
|Total number of errors encountered during backup
|Gauge

|velero_backup_items_total
|Total number of items backed up
|Gauge

|velero_backup_last_status

> |Last status of the backup. A value of 1 is success, 0

Reviewer 2: ```suggestion
|Last status of the backup. A value of 1 is success, 0.
```

I think, since it's mixed here?

|Total number of validation failed backups
|Counter

|velero_backup_warning_total
|Total number of warned backups
|Counter

|velero_csi_snapshot_attempt_total
|Total number of CSI attempted volume snapshots
|Counter

|velero_csi_snapshot_failure_total
|Total number of CSI failed volume snapshots
|Counter

|velero_csi_snapshot_success_total
|Total number of CSI successful volume snapshots
|Counter

|velero_restore_attempt_total
|Total number of attempted restores
|Counter

|velero_restore_failed_total
|Total number of failed restores
|Counter

|velero_restore_partial_failure_total
|Total number of partially failed restores
|Counter

|velero_restore_success_total
|Total number of successful restores
|Counter

|velero_restore_total
|Current number of existent restores
|Gauge

|velero_restore_validation_failed_total
|Total number of failed restores failing validations
|Counter

|velero_volume_snapshot_attempt_total
|Total number of attempted volume snapshots
|Counter

|velero_volume_snapshot_failure_total
|Total number of failed volume snapshots
|Counter

|velero_volume_snapshot_success_total
|Total number of successful volume snapshots
|Counter

|===

> [id="Viewing-metrics-OpenShift-Observe-UI_{context}"]
> ==  Viewing metrics using the OpenShift Observe UI

Reviewer 2: ```suggestion
[id="viewing-metrics-observe-ui_{context}"]
==  Viewing metrics using the Observe UI
```

Lowercase, and does it need to say "OpenShift"?

Also, break this out into a procedure module.

> The metrics can be viewed in the OpenShift Administrator or Developer view, which must have access to the `openshift-adp project`.

Reviewer 2: ```suggestion
You can view metrics in the {product-title} web console from the *Administrator* or *Developer* perspective, which must have access to the `openshift-adp` project.
```

.Procedure

> . Select *Observe* and then select *Metrics*.
> .. In the *Developer* view select *Custom query*, or click on the *Show PromQL* link, type the query and click *Enter*.
> .. In the *Administrator* view type the Expression in the text field and select *Run Queries*.

Reviewer 2: Technically since this is a one-step procedure (even though it has substeps), it should be an asterisk instead of `.` for the first step. There seemed to be a lot of steps jammed together though past that, so I took a crack at sorting it out, as a suggestion, though it does get a little wonky with the back-and-forth unordered/ordereds:

```suggestion
* Navigate to the *Observe* -> *Metrics* page:

** If you are using the *Developer* perspective, follow these steps:
.. Select *Custom query*, or click on the *Show PromQL* link.
.. Type the query and click *Enter*.

** If you are using the *Administrator* perspective, type the expression in the text field and select *Run Queries*.
```

Side question: is it really "clicking" `Enter`, or does it mean pressing the Enter key? I didn't see an **Enter** button in the screenshot, so I wondered.

+
.OADP metrics query

> image::oadp-metrics-query.png[OADP metrics query]

Reviewer 2: And speaking of the screenshot here, I noticed it was showing the **Developer** perspective, but with the `+` where it is, it makes it look like it's attached to the **Administrator** perspective step. I'd suggest either moving the screenshot up to the **Developer** step, or just remove the `+` here and let it kind of float after the procedure list is technically finished.

With enabled User Workload Monitoring it is possible to configure and use any Prometheus-compatible third-party UI, such as Grafana to visualize Velero metrics.

Monitoring metrics requires enabling monitoring for the user-defined projects and creating a ServiceMonitor resource to scrape those metrics from the already enabled OADP service endpoint that resides in the `openshift-adp` namespace.

.Prerequisites
* You created an OADP operator, a credentials secret, and a DataProtectionApplication (DPA) custom resource (CR).
* Your shell has the `oc` CLI command available.
* You have Administrator level access to the OpenShift cluster.

.Procedure

. Edit the `cluster-monitoring-config` ConfigMap object in the `openshift-monitoring` namespace and add or enable the `enableUserWorkload` option under `data/config.yaml`.
+
[source,terminal]
----

> $ oc edit configmap cluster-monitoring-config -n openshift-monitoring

Author: @mpryc 
I added to the prerequisites.
- You have created a cluster monitoring config map (with a xref).

Is that OK?

The OADP Operator leverages an OpenShift xref:../../monitoring/enabling-monitoring-for-user-defined-projects.adoc#enabling-monitoring-for-user-defined-projects_enabling-monitoring-for-user-defined-projects[User Workload Monitoring] provided by the OpenShift Monitoring Stack for retrieving metrics from the Velero service endpoint. The monitoring stack allows creating user-defined Alerting Rules or querying metrics, using the OpenShift Metrics query front-end.

With enabled User Workload Monitoring it is possible to configure and use any Prometheus-compatible third-party UI, such as Grafana to visualize Velero metrics.

[id="OADP-monitoring-setup_{context}"]
== OADP monitoring setup
Monitoring metrics requires enabling monitoring for the user-defined projects and creating a ServiceMonitor resource to scrape those metrics from the already enabled OADP service endpoint that resides in the `openshift-adp` namespace.

> [NOTE]

Reviewer: 1. I am not sold on adding this NOTE in the official documentation. The configuration files in the source code of the project are mostly for the development and users who follow docs within the project source code and not the official OCP docs.

2.  The instructions below also enables the user workload monitoring, however on the line 11 there of this file there is link to already existing procedure to enable user workload monitoring and lives in the *same* "home, so I wonder if the entire "OADP monitoring setup" should be just a paragraph here with first sentence:
   ```
Monitoring metrics requires enabling monitoring for the user-defined projects
and creating a ServiceMonitor resource to scrape those metrics from the
already enabled OADP service endpoint that resides in the openshift-adp namespace.
```

Plus some Prerequisites, with Procedure as a link to the:
 ```
xref:../../monitoring/monitoring-overview.adoc#enabling-monitoring-for-user-defined-projects_enabling-monitoring-for-user-defined-projects[Enabling monitoring for user-defined projects]
```

The source code contains that procedure as it's outside of the Red Hat OpenShift Documentation, however here I see duplication.

NOTE, that maybe it's better to keep it here as the few lines gives end to end instructions that are crafted for OADP and the user won't need to jump between different documentation parts and it's just few lines of docs that duplicates the instructions.

Author: @mpryc 
Thank you for your feedback. I made some changes based on your feedback.

====
The configuration files that enable User Workload Monitoring can be found https://github.com/openshift/oadp-operator/tree/master/docs/examples/manifests/user_monitoring[here], however they will override any previous configuration that could have been on the cluster, so use them with caution. They are named with a number prefix, which represents the order in which they should be applied.
====

.Prerequisites
* You created an OADP operator, a credentials secret, and a DataProtectionApplication (DPA) custom resource (CR).
* Your shell has the `oc` CLI command available.
* You have Administrator level access to the OpenShift cluster.

.Procedure

. Edit the `cluster-monitoring-config` ConfigMap object in the `openshift-monitoring` namespace and add or enable the `enableUserWorkload` option under `data/config.yaml`.
+
[source,terminal]
----
$ oc edit configmap cluster-monitoring-config -n openshift-monitoring
----
[source,yaml]
+
----
apiVersion: v1
data:
  config.yaml: |
    enableUserWorkload: true  # Add this option or set to true
kind: ConfigMap
metadata:
# [...]
----
+

> . Wait a short period of time to verify the User Workload Monitoring Setup by checking if the following components are up and running in the `openshift-user-workload-monitoring` namespace.

Author: What is a short period of time? Can we write in minutes, for example several minutes?

Reviewer: It really depends on the cluster, but my estimate is somewhere between few seconds up to 2-3 minutes. I was trying to find a single command which will watch a particular resource to be ready, however we have here multiple of those so it's not that simple.

See other part of the documentation which is pretty similar and that wording in point number 4:

https://62710--docspreview.netlify.app/openshift-enterprise/latest/monitoring/enabling-monitoring-for-user-defined-projects.html#enabling-monitoring-for-user-defined-projects_enabling-monitoring-for-user-defined-projects
